### PR TITLE
Moveit simple Issue #8

### PR DESCRIPTION
**Pull request by MauleshSTrivedi**
Added functionality for dynamically re-configuring the trajectory executing speed using rosparam_handler.

### CODE REVIEW

**Comment by shaun-edwards** [comment]
Relies on #20.

File: moveit_simple/CMakeLists.txt [hunk 1]
```diff
@@ -2,6 +2,8 @@ cmake_minimum_required(VERSION 2.8.3)
 project(moveit_simple)
 find_package(catkin REQUIRED COMPONENTS
   roscpp
+  rosparam_handler
+  dynamic_reconfigure
```
Comment (shaun-edwards) [inline moveit_simple/CMakeLists.txt]: Do you need to depend on both rosparam_handler and dynamic_reconfigure?  One should include the other.

File: moveit_simple/include/moveit_simple/moveit_simple.h [hunk 2]
```diff
@@ -388,19 +473,54 @@ class OnlineRobot : public Robot
...
+   * prior calls to "addTrajPoint".
+   * @param goal - Joint trajectory goal
+   * @param collision_check - bool to turn check for collision on\off
+   * @throws <moveit_simple::ExecutionFailureException> (Execution failure)
+   * @throws <moveit_simple::IKFailException> (Conversion to joint trajectory failed)
+   * @throws <std::invalid_argument> (Trajectory "traj_name" not found)
+   * @throws <moveit_simple::CollisionDetected> (One of interpolated point is
+   * in Collision with itself or environment)
+   */
+  void planOnly(const std::string traj_name,
```
Comment (shaun-edwards) [inline moveit_simple/include/moveit_simple/moveit_simple.h]: Change name to `plan`....the Only modify does not add anything.

File: moveit_simple/include/moveit_simple/moveit_simple.h [hunk 3]
```diff
@@ -388,19 +473,54 @@ class OnlineRobot : public Robot
...
+   * @return
+   */
+  void setSpeedModifier(double speed_modifier);
+
+  /**
+   * @brief setSpeedModifier - Getter method for the execution speed modifier of the 
+   * execute method.
+   * @return speed_modifier_
+   */
+  double getSpeedModifier(void);
```
Comment (shaun-edwards) [inline moveit_simple/include/moveit_simple/moveit_simple.h]: Should be a `const` method.

File: moveit_simple/include/moveit_simple/moveit_simple.h [hunk 4]
```diff
@@ -410,6 +530,14 @@ class OnlineRobot : public Robot
 
   // ROS objects
   ros::Subscriber j_state_sub_;
+
+  // Dynamic Reconfigure
+  double speed_modifier_;
```
Comment (shaun-edwards) [inline moveit_simple/include/moveit_simple/moveit_simple.h]: I know this is based on my original code, but the terminology and sematics of this are bad.  Can we change it to be `speed_percent` with values from 0->1.0?

**Review by shaun-edwards — CHANGES_REQUESTED** [review]
Please rebase after #20 has been merged.  Also, you are using dynamic reconfigure and the param handler...you should only be using the later.  Please remove references to dynamic reconfigure as these are not needed.

**Comment by shaun-edwards** [comment]
There's too many changes from #20...I will review again once #20 has been merged and this branch has been rebased on indigo-devel.

File: moveit_simple/CMakeLists.txt (hunk 1 above)
Comment (MauleshSTrivedi) [reply moveit_simple/CMakeLists.txt]: I initially ran the tests with just dynamic_reconfigure but later ported to rosparam_handler.. I will remove the necessary dependencies.

File: moveit_simple/include/moveit_simple/moveit_simple.h (hunk 2 above)
Comment (MauleshSTrivedi) [reply moveit_simple/include/moveit_simple/moveit_simple.h]: ok

File: moveit_simple/include/moveit_simple/moveit_simple.h (hunk 4 above)
Comment (MauleshSTrivedi) [reply moveit_simple/include/moveit_simple/moveit_simple.h]: ok, sure, I can do that.

File: moveit_simple/CMakeLists.txt (hunk 1 above)
Comment (MauleshSTrivedi) [reply moveit_simple/CMakeLists.txt]: yes, so apparently dynamic_reconfigure needs to be explicitly added otherwise we get dependency issues.

**Comment by MauleshSTrivedi** [comment]
@shaun-edwards Could you please review this. Thanks a lot.

File: moveit_simple/include/moveit_simple/moveit_simple.h [hunk 5]
```diff
@@ -465,19 +468,71 @@ class OnlineRobot : public Robot
...
+   * in Collision with itself or environment)
+   */
+  void executeKnownPlan(const std::string traj_name, 
+                   control_msgs::FollowJointTrajectoryGoal & goal,
+                   bool collision_check = false);
+
+
+  /**
+   * @brief plan out a given trajectory
+   * @param traj_name - name of trajectory to be executed (must be filled with
```
Comment (shaun-edwards) [inline moveit_simple/include/moveit_simple/moveit_simple.h]: Why is the trajectory name required?

File: moveit_simple/include/moveit_simple/moveit_simple.h [hunk 6]
```diff
@@ -465,19 +468,71 @@ class OnlineRobot : public Robot
...
+   * prior calls to "addTrajPoint".
+   * @param goal - Joint trajectory goal
+   * @param collision_check - bool to turn check for collision on\off
+   * @throws <moveit_simple::ExecutionFailureException> (Execution failure)
+   * @throws <moveit_simple::IKFailException> (Conversion to joint trajectory failed)
```
Comment (shaun-edwards) [inline moveit_simple/include/moveit_simple/moveit_simple.h]: Can this still happen if we pass in a joint trajectory?

File: moveit_simple/include/moveit_simple/moveit_simple.h [hunk 7]
```diff
@@ -465,19 +468,71 @@ class OnlineRobot : public Robot
...
+  void executeKnownPlan(const std::string traj_name, 
+                   control_msgs::FollowJointTrajectoryGoal & goal,
+                   bool collision_check = false);
+
+
+  /**
+   * @brief plan out a given trajectory
+   * @param traj_name - name of trajectory to be executed (must be filled with
+   * prior calls to "addTrajPoint".
+   * @param goal - Joint trajectory goal
```
Comment (shaun-edwards) [inline moveit_simple/include/moveit_simple/moveit_simple.h]: Using this type leaks ROS types into moveit_simple.  Can we use out own data types instead?

File: moveit_simple/include/moveit_simple/moveit_simple.h [hunk 8]
```diff
@@ -465,19 +468,71 @@ class OnlineRobot : public Robot
...
+   * @param collision_check - bool to turn check for collision on\off
+   * @throws <moveit_simple::ExecutionFailureException> (Execution failure)
+   * @throws <moveit_simple::IKFailException> (Conversion to joint trajectory failed)
+   * @throws <std::invalid_argument> (Trajectory "traj_name" not found)
+   * @throws <moveit_simple::CollisionDetected> (One of interpolated point is
+   * in Collision with itself or environment)
+   */
+  void plan(const std::string traj_name,
```
Comment (shaun-edwards) [inline moveit_simple/include/moveit_simple/moveit_simple.h]: Why not return the plan?

File: moveit_simple/include/moveit_simple/moveit_simple.h [hunk 9]
```diff
@@ -465,19 +468,71 @@ class OnlineRobot : public Robot
...
+   * @param collision_check - bool to turn check for collision on\off
+   * @throws <moveit_simple::ExecutionFailureException> (Execution failure)
+   * @throws <moveit_simple::IKFailException> (Conversion to joint trajectory failed)
+   * @throws <std::invalid_argument> (Trajectory "traj_name" not found)
+   * @throws <moveit_simple::CollisionDetected> (One of interpolated point is
+   * in Collision with itself or environment)
+   */
+  void plan(const std::string traj_name, 
+          control_msgs::FollowJointTrajectoryGoal & goal,
+          ros::Duration & traj_time,
```
Comment (shaun-edwards) [inline moveit_simple/include/moveit_simple/moveit_simple.h]: Why is this required?

File: moveit_simple/include/moveit_simple/moveit_simple.h [hunk 10]
```diff
@@ -465,19 +468,71 @@ class OnlineRobot : public Robot
...
+   * @param goal - Joint trajectory goal
+   * @param collision_check - bool to turn check for collision on\off
+   * @throws <moveit_simple::ExecutionFailureException> (Execution failure)
+   * @throws <moveit_simple::IKFailException> (Conversion to joint trajectory failed)
+   * @throws <std::invalid_argument> (Trajectory "traj_name" not found)
+   * @throws <moveit_simple::CollisionDetected> (One of interpolated point is
+   * in Collision with itself or environment)
+   */
+  void plan(const std::string traj_name, 
+          control_msgs::FollowJointTrajectoryGoal & goal,
```
Comment (shaun-edwards) [inline moveit_simple/include/moveit_simple/moveit_simple.h]: See comment above about leaking ROS types.

File: moveit_simple/include/moveit_simple/moveit_simple.h [hunk 11]
```diff
@@ -465,19 +468,71 @@ class OnlineRobot : public Robot
...
   */
   virtual std::vector<double> getJointState(void) const;
 
+  /**
+   * @brief setSpeedModifier - Setter method for the execution speed modifier of the 
+   * execute method.
+   * @param speed_modifier
+   * @return
+   */
+  void setSpeedModifier(double speed_modifier);
```
Comment (shaun-edwards) [inline moveit_simple/include/moveit_simple/moveit_simple.h]: The function name and member variable should have the same name (or about).  This makes it easier to correlate dynamic reconfigure parameters with member variables.

File: moveit_simple/include/moveit_simple/moveit_simple.h [hunk 12]
```diff
@@ -465,19 +468,71 @@ class OnlineRobot : public Robot
...
+   * prior calls to "addTrajPoint".
+   * @param goal - Joint trajectory goal which is a known 'Plan'
+   * @param collision_check - bool to turn check for collision on\off
+   * @throws <moveit_simple::ExecutionFailureException> (Execution failure)
+   * @throws <moveit_simple::IKFailException> (Conversion to joint trajectory failed)
+   * @throws <std::invalid_argument> (Trajectory "traj_name" not found)
+   * @throws <moveit_simple::CollisionDetected> (One of interpolated point is
+   * in Collision with itself or environment)
+   */
+  void executeKnownPlan(const std::string traj_name,
```
Comment (shaun-edwards) [inline moveit_simple/include/moveit_simple/moveit_simple.h]: As opposed to an unknown plan?  The method signature should be different for this method, allowing you to use the same `execute` name.

File: moveit_simple/include/moveit_simple/moveit_simple.h [hunk 13]
```diff
@@ -465,19 +468,71 @@ class OnlineRobot : public Robot
...
+   * @return
+   */
+  void setSpeedModifier(double speed_modifier);
+
+  /**
+   * @brief setSpeedModifier - Getter method for the execution speed modifier of the 
+   * execute method.
+   * @return speed_modifier_
+   */
+  double getSpeedModifier(void) const;
```
Comment (shaun-edwards) [inline moveit_simple/include/moveit_simple/moveit_simple.h]: See comment above about naming.

File: moveit_simple/include/moveit_simple/moveit_simple.h (hunk 8 above)
Comment (shaun-edwards) [inline moveit_simple/include/moveit_simple/moveit_simple.h]: Should the `plan` method be a part of `OnlineRobot` or the base class?

File: moveit_simple/include/moveit_simple/moveit_simple.h (hunk 11 above)
Comment (shaun-edwards) [inline moveit_simple/include/moveit_simple/moveit_simple.h]: If you move Plan into the base class, then this member method and dynamic reconfigure should be moved to the base class.

File: moveit_simple/src/moveit_simple.cpp [hunk 14]
```diff
@@ -640,6 +720,55 @@ void OnlineRobot::execute(const std::string traj_name, bool collision_check)
...
+
+
+void OnlineRobot::plan(const std::string traj_name, 
+                       control_msgs::FollowJointTrajectoryGoal & goal,
+                       ros::Duration & traj_time,
+                       bool collision_check)
+{
+  std::lock_guard<std::recursive_mutex> guard(m_);
+
+  const double TIMEOUT_SCALE = 1.25;  //scales time to wait for action timeout.
```
Comment (shaun-edwards) [inline moveit_simple/src/moveit_simple.cpp]: This variable is no longer needed.

File: moveit_simple/src/moveit_simple.cpp [hunk 15]
```diff
@@ -659,6 +788,46 @@ void OnlineRobot::execute(const std::string traj_name, bool collision_check)
 
 
 
+void OnlineRobot::reconfigureRequest(moveit_simple_dynamic_reconfigure_Config &config, uint32_t level)
+{
+  params_.fromConfig(config);
+  if (params_.speed_percent > 0.0)
```
Comment (shaun-edwards) [inline moveit_simple/src/moveit_simple.cpp]: What do you do if this isn't true.

File: moveit_simple/src/moveit_simple.cpp [hunk 16]
```diff
@@ -659,6 +788,46 @@ void OnlineRobot::execute(const std::string traj_name, bool collision_check)
...
+  {
+    setSpeedModifier(params_.speed_percent);
+  }
+}
+
+
+
+void OnlineRobot::setSpeedModifier(double speed_percent)
+{
+  if (speed_percent <= 1.0 && speed_percent > 0.0)
```
Comment (shaun-edwards) [inline moveit_simple/src/moveit_simple.cpp]: You should use constants.

File: moveit_simple/src/moveit_simple.cpp [hunk 17]
```diff
@@ -659,6 +788,46 @@ void OnlineRobot::execute(const std::string traj_name, bool collision_check)
...
+  {
+    speed_modifier_ = speed_percent;
+  }
+  else if((speed_percent > 1.0))  
+  {
+    speed_modifier_ = max_speed;
```
Comment (shaun-edwards) [inline moveit_simple/src/moveit_simple.cpp]: Don't declare constant member varriables unless they are used in multiple methods.  Otherwise just create them in the method in which they are used.

File: moveit_simple/src/moveit_simple.cpp [hunk 18]
```diff
@@ -659,6 +788,46 @@ void OnlineRobot::execute(const std::string traj_name, bool collision_check)
...
+    ROS_WARN_STREAM("Clamping Speed from " << speed_modifier_ <<
+                    " to max_speed: " << max_speed);
+  }
+  else if((speed_percent <= 0.0))  
+  {
+    speed_modifier_ = min_speed;
```
Comment (shaun-edwards) [inline moveit_simple/src/moveit_simple.cpp]: See comment above.

File: moveit_simple/src/moveit_simple.cpp [hunk 19]
```diff
@@ -659,6 +788,46 @@ void OnlineRobot::execute(const std::string traj_name, bool collision_check)
...
+  {
+    speed_modifier_ = speed_percent;
+  }
+  else if((speed_percent > 1.0))  
+  {
+    speed_modifier_ = max_speed;
+    ROS_WARN_STREAM("Clamping Speed from " << speed_modifier_ <<
+                    " to max_speed: " << max_speed);
+  }
+  else if((speed_percent <= 0.0))
```
Comment (shaun-edwards) [inline moveit_simple/src/moveit_simple.cpp]: You should always have an `else`....of if you can't get to an else, remove the last `else if` condition.

**Review by shaun-edwards — CHANGES_REQUESTED** [review]
Please address my comments and resubmit.  I didn't tag all the changes...I will wait for you to update the code.

File: moveit_simple/include/moveit_simple/moveit_simple.h (hunk 12 above)
Comment (MauleshSTrivedi) [reply moveit_simple/include/moveit_simple/moveit_simple.h]: ok, I've changed the signature.

File: moveit_simple/include/moveit_simple/moveit_simple.h (hunk 5 above)
Comment (MauleshSTrivedi) [reply moveit_simple/include/moveit_simple/moveit_simple.h]: Its needed in the call for toJointTrajectory(...) inside plan(...).

File: moveit_simple/include/moveit_simple/moveit_simple.h (hunk 6 above)
Comment (MauleshSTrivedi) [reply moveit_simple/include/moveit_simple/moveit_simple.h]: No it cannot. But Plan returns a jointTrajectory.. the execute(...) function overload takes a jointTrajectory input and executes said traj. And yes, it will NOT throw an IKFailException. Please check out the documentation for that execute(...) function for more details on the exceptions it throws.

File: moveit_simple/include/moveit_simple/moveit_simple.h (hunk 8 above)
Comment (MauleshSTrivedi) [reply moveit_simple/include/moveit_simple/moveit_simple.h]: Moved it to base class.

File: moveit_simple/include/moveit_simple/moveit_simple.h (hunk 9 above)
Comment (MauleshSTrivedi) [reply moveit_simple/include/moveit_simple/moveit_simple.h]: It was used for some other reason but I forgot to deprecate it. It has been removed now.

File: moveit_simple/include/moveit_simple/moveit_simple.h (hunk 11 above)
Comment (MauleshSTrivedi) [reply moveit_simple/include/moveit_simple/moveit_simple.h]: Done.

File: moveit_simple/src/moveit_simple.cpp (hunk 15 above)
Comment (MauleshSTrivedi) [reply moveit_simple/src/moveit_simple.cpp]: Modified it to reflect this case.

File: moveit_simple/include/moveit_simple/moveit_simple.h (hunk 8 above)
Comment (MauleshSTrivedi) [reply moveit_simple/include/moveit_simple/moveit_simple.h]: I was technically sending a reference as an argument. I've modified it so that it returns the plan now.

Comment (MauleshSTrivedi) [reply moveit_simple/include/moveit_simple/moveit_simple.h]: Returning a complex data structure like a plan is not something I would prefer. Instead, we pass an empty plan (type: control_msgs::FollowJointTrajectoryGoal) as a reference argument to the Plan(...) function. This populates the joint trajectory plan which can later be used in the execute(...) function.

**Comment by MauleshSTrivedi** [comment]
@shaun-edwards Could you please review these final changes. Thanks.

File: moveit_simple/include/moveit_simple/moveit_simple.h (hunk 8 above)
Comment (shaun-edwards) [reply moveit_simple/include/moveit_simple/moveit_simple.h]: Is there a reason not to return the data type? The use of this method will read "better" this way?

Comment (MauleshSTrivedi) [reply moveit_simple/include/moveit_simple/moveit_simple.h]: Yeah, it does read better that way. I guess the main reason is that I usually pass complex data types as references for efficiency reasons. But in this case you are right. The method is going to be exposed at the user's end and not at the developer's end. That supersedes my reason for efficiency. I will make the changes.

Comment (gavanderhoorn) [reply moveit_simple/include/moveit_simple/moveit_simple.h]: @MauleshSTrivedi: wouldn't RVO take care of this?

Also: I would suggest to avoid this sort of premature optimisation. Methods/functions should return proper types, it makes the intent clear. If/when it turns out that returning 'complex types' (debatable for the plan struct) is a bottleneck, *then* it could make sense to pass in reference objects from caller to callee.

Comment (MauleshSTrivedi) [reply moveit_simple/include/moveit_simple/moveit_simple.h]: yeah actually I forgot about RVO.. Instinctively I usually go for references for these types but your'e right, RVO should take care of this.. :)

**Comment by MauleshSTrivedi** [comment]
@shaun-edwards Hey Shaun, updated the PR.. lemme know what you think. :)

File: moveit_simple/include/moveit_simple/moveit_simple.h [hunk 20]
```diff
@@ -71,6 +74,8 @@ struct TrajectoryPointInfo {
 };
 typedef std::vector<TrajectoryPointInfo> TrajectoryInfo;
 
+typedef control_msgs::FollowJointTrajectoryGoal JointTrajectoryType;
```
Comment (shaun-edwards) [inline moveit_simple/include/moveit_simple/moveit_simple.h]: This should be a moveit_simple type and not a ROS type (as we discussed).

File: moveit_simple/src/moveit_simple.cpp [hunk 21]
```diff
@@ -611,6 +619,50 @@ void Robot::clearTrajectory(const::std::string traj_name)
...
+    goal.trajectory.joint_names = joint_group_->getVariableNames();
+    try
+    {
+      if ( toJointTrajectory(traj_name, goal.trajectory.points, collision_check) )
+      {
+
+        // Modify the speed of execution for the trajectory based off of the speed_modifier_
+        for (std::size_t i = 0; i < goal.trajectory.points.size(); i++)
+        {
+          goal.trajectory.points[i].time_from_start *= (1.0/speed_modifier_);
```
Comment (shaun-edwards) [inline moveit_simple/src/moveit_simple.cpp]: The trajectory scaling by the `speed_modifier_` should either occur in the `plan` method or the `execute` method, but not both.  I'd recommend the `plan` method because that was the whole reason for pulling the `speed_modifier_` into the `Robot` base type.

File: moveit_simple/src/moveit_simple.cpp [hunk 22]
```diff
@@ -611,6 +619,50 @@ void Robot::clearTrajectory(const::std::string traj_name)
...
+  else
+  {
+    ROS_ERROR_STREAM("Trajectoy [" << traj_name << "] not found");
+    throw std::invalid_argument("No trajectory found named " + traj_name);
+  }
+}
+
+
+
 void OnlineRobot::execute(const std::string traj_name, bool collision_check)
```
Comment (shaun-edwards) [inline moveit_simple/src/moveit_simple.cpp]: Why doesn't this method call `plan` followed by `execute` with the planned trajectory?

File: moveit_simple/src/moveit_simple.cpp [hunk 23]
```diff
@@ -619,46 +671,157 @@ void OnlineRobot::execute(const std::string traj_name, bool collision_check)
...
   {
-    ROS_ERROR_STREAM("Trajectoy " << traj_name << " not found");
+    ROS_ERROR_STREAM("Trajectoy [" << traj_name << "] not found");
     throw std::invalid_argument("No trajectory found named " + traj_name);
   }
 }
 
 
 
+void OnlineRobot::execute(JointTrajectoryType & goal,
```
Comment (shaun-edwards) [inline moveit_simple/src/moveit_simple.cpp]: The method should simply check all points for collision, then populate the action goal and send.  If there is any collision, then the method should throw.  The structure below does not have to be that complex.

File: moveit_simple/src/moveit_simple.cpp [hunk 24]
```diff
@@ -619,46 +671,157 @@ void OnlineRobot::execute(const std::string traj_name, bool collision_check)
...
+  }
+  else if ((speed_modifier > 1.0))  
+  {
+    speed_modifier_ = 1.0;
+    ROS_WARN_STREAM("Clamping Speed from " << speed_modifier_ <<
+                    " to max_speed: [1.0]");
+  }
+  else
+  {
+    speed_modifier_ = 0.1;
```
Comment (shaun-edwards) [inline moveit_simple/src/moveit_simple.cpp]: I recommend a lower limit....like 0.01

**Review by shaun-edwards — CHANGES_REQUESTED** [review]

File: moveit_simple/include/moveit_simple/moveit_simple.h (hunk 20 above)
Comment (MauleshSTrivedi) [reply moveit_simple/include/moveit_simple/moveit_simple.h]: done

File: moveit_simple/src/moveit_simple.cpp (hunk 24 above)
Comment (MauleshSTrivedi) [reply moveit_simple/src/moveit_simple.cpp]: I have changed this param, but I would recommend against this.. the simple reason being that a speed of 0.01 is soo soo slow that, more often than not, the program crashes due to some problem in max timeout or thread or something similar.

**Comment by MauleshSTrivedi** [comment]
@shaun-edwards Could you have a look at this update please. Thanks.

File: moveit_simple/include/moveit_simple/moveit_simple.h [hunk 25]
```diff
@@ -318,6 +341,12 @@ class Robot
                                          const std::string& frame_in,
                                          const std::string& frame_out) const;
 
+  std::vector<moveit_simple::JointTrajectoryPoint> 
+    trajectoryTypeConversionFromROSToMoveItSimple(std::vector<trajectory_msgs::JointTrajectoryPoint> & ROS_joint_trajectory_points)  const;
```
Comment (shaun-edwards) [inline moveit_simple/include/moveit_simple/moveit_simple.h]: Way too long.  The return and parameter types make this obvious.  How about `toJointTrajectoryPoint` and `toFollowJointTrajectoryGoal`.  Because parameters can be different (overrides), but return types cannot, this ensures you will be able to support future conversions.

File: moveit_simple/include/moveit_simple/moveit_simple.h [hunk 26]
```diff
@@ -383,6 +412,10 @@ class Robot
   bool isReachable(std::unique_ptr<TrajectoryPoint> & point, double timeout,
                    std::vector<double> joint_seed = std::vector<double>() ) const;
 
+  bool jointTrajectoryCollisionCheck(control_msgs::FollowJointTrajectoryGoal & goal,
```
Comment (shaun-edwards) [inline moveit_simple/include/moveit_simple/moveit_simple.h]: This could be shorter...again the parameters indicate this works on a joint trajectory...so no need for an extra long name.

File: moveit_simple/include/moveit_simple/moveit_simple.h [hunk 27]
```diff
@@ -406,9 +439,22 @@ class Robot
...
                       const std::vector<double> jp2) const;
 
+  /**
+   * @brief toJointTrajPtMsg - Converts native joint point (vector + time) to ROS joint
+   * trajectory point message type.
+   * @param joint_point
+   * @param time
+   * @return
+   */
+  static trajectory_msgs::JointTrajectoryPoint toJointTrajPtMsg(
```
Comment (shaun-edwards) [inline moveit_simple/include/moveit_simple/moveit_simple.h:446]: Good naming!

File: moveit_simple/include/moveit_simple/moveit_simple.h [hunk 28]
```diff
@@ -406,9 +439,22 @@ class Robot
...
+      const std::vector<double> & joint_point, double time);
```
Comment (shaun-edwards) [inline moveit_simple/include/moveit_simple/moveit_simple.h:447]: What is the `time` parameter used for?

File: moveit_simple/src/moveit_simple.cpp [hunk 29]
```diff
@@ -611,54 +619,233 @@ void Robot::clearTrajectory(const::std::string traj_name)
...
 {
   std::lock_guard<std::recursive_mutex> guard(m_);
 
-  const double TIMEOUT_SCALE = 1.25;  //scales time to wait for action timeout.
-  bool success = false;
   if ( traj_info_map_.count(traj_name) )
   {
-    control_msgs::FollowJointTrajectoryGoal goal;
-    goal.trajectory.joint_names = joint_group_->getVariableNames();
+  	std::vector<trajectory_msgs::JointTrajectoryPoint> ROS_trajectory_points;
```
Comment (shaun-edwards) [inline moveit_simple/src/moveit_simple.cpp:629]: Correct spacing.

Comment (shaun-edwards) [reply moveit_simple/src/moveit_simple.cpp:629]: naming should be `lower_case`

File: moveit_simple/src/moveit_simple.cpp [hunk 30]
```diff
@@ -611,54 +619,233 @@ void Robot::clearTrajectory(const::std::string traj_name)
...
+
+
+void OnlineRobot::execute(std::vector<moveit_simple::JointTrajectoryPoint> & joint_trajectory_points,
+                          bool collision_check)
+{
+  std::lock_guard<std::recursive_mutex> guard(m_);
+
+  const double TIMEOUT_SCALE = 1.25;  //scales time to wait for action timeout.
+  
+  int collision_points = 0;
```
Comment (shaun-edwards) [inline moveit_simple/src/moveit_simple.cpp]: Is this still used?

File: moveit_simple/include/moveit_simple/moveit_simple.h [hunk 31]
```diff
@@ -383,6 +412,10 @@ class Robot
   bool isReachable(std::unique_ptr<TrajectoryPoint> & point, double timeout,
                    std::vector<double> joint_seed = std::vector<double>() ) const;
 
+  bool jointTrajectoryCollisionCheck(control_msgs::FollowJointTrajectoryGoal & goal,
+                                     int & collision_points,
```
Comment (shaun-edwards) [inline moveit_simple/include/moveit_simple/moveit_simple.h]: This parameter seems redundant with the return value.  Could you just return `0` if not collisions are found?

File: moveit_simple/src/moveit_simple.cpp [hunk 32]
```diff
@@ -611,54 +619,233 @@ void Robot::clearTrajectory(const::std::string traj_name)
...
+
+  	throw CollisionDetected("Collision detected while interpolating ");
+  }
+}
+
+
+
+std::vector<moveit_simple::JointTrajectoryPoint> Robot::trajectoryTypeConversionFromROSToMoveItSimple(std::vector<trajectory_msgs::JointTrajectoryPoint> & ROS_joint_trajectory_points)  const
+{
+	std::vector<moveit_simple::JointTrajectoryPoint> goal;
```
Comment (shaun-edwards) [inline moveit_simple/src/moveit_simple.cpp]: Don't change, but since you know the size of the trajectory you are converting, you may consider calling `reserve` ahead of time for better performance.

File: moveit_simple/src/moveit_simple.cpp [hunk 33]
```diff
@@ -611,54 +619,233 @@ void Robot::clearTrajectory(const::std::string traj_name)
...
+
+bool Robot::jointTrajectoryCollisionCheck(control_msgs::FollowJointTrajectoryGoal & goal,
+										  int & collision_count,
+										  bool collision_check)	
+{
+	bool isStatecolliding = false;
+	collision_count = 0;
+
+	if(collision_check)
+	{
```
Comment (shaun-edwards) [inline moveit_simple/src/moveit_simple.cpp]: Correct spacing.  `tabs` should be `2 spaces`.

**Review by shaun-edwards — CHANGES_REQUESTED** [review]

File: moveit_simple/include/moveit_simple/moveit_simple.h (hunk 31 above)
Comment (MauleshSTrivedi) [reply moveit_simple/include/moveit_simple/moveit_simple.h]: ok

File: moveit_simple/include/moveit_simple/moveit_simple.h (hunk 28 above)
Comment (MauleshSTrivedi) [reply moveit_simple/include/moveit_simple/moveit_simple.h:447]: For time from start.

File: moveit_simple/src/moveit_simple.cpp (hunk 30 above)
Comment (MauleshSTrivedi) [reply moveit_simple/src/moveit_simple.cpp]: yes, but only in this method.

**Comment by MauleshSTrivedi** [comment]
@shaun-edwards Could you review these final changes for Issue #8 (Anytime after 15 mins from now -- Travis is still building). Thanks.

**Review by shaun-edwards — APPROVED** [review]
Looks good!  Thanks for the effort and patience.

**Comment by MauleshSTrivedi** [comment]
Thanks a lot Shaun. Anytime. :)
